llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -8121,6 +8124,21 @@ llvm::GlobalValue::LinkageTypes SPIRVToLLVM::transLinkageType(const SPIRVValue *
   }
 }
 
+llvm::Function *SPIRVToLLVM::createLibraryEntryFunc() {
+  auto builder = getBuilder();
+  FunctionType *funcTy = FunctionType::get(builder->getVoidTy(), {}, false);
+  auto func = Function::Create(funcTy, GlobalValue::ExternalLinkage, "libraryEntry", m_m);
+  BasicBlock *entryBlock = BasicBlock::Create(*m_context, "", func);
+  builder->SetInsertPoint(entryBlock);
+  builder->CreateRetVoid();
+  std::vector<Metadata *> execModelMDs;
+  execModelMDs.push_back(ConstantAsMetadata::get(ConstantInt::get(builder->getInt32Ty(), ExecutionModelGLCompute)));
+  auto execModelMdNode = MDNode::get(*m_context, execModelMDs);
+  func->addMetadata(gSPIRVMD::ExecutionModel, *execModelMdNode);
+  func->setDLLStorageClass(GlobalValue::DLLExportStorageClass);
```

Comment:
IIRC @trenouf once said that `dllexport` is used to mark shader entry points (i.e. functions that are called by hardware in contrast to functions called by software), so I think this should not be marked as `dllexport` (but let’s wait for him to confirm).

Reply:
spirv with capability linkage is library spirv,  hence it can not be the module with hardware entry function.
however we need to get the module information i.e. module shader stage, from entry function.
so i mark that dummy function as DLLExportStorageClass

Reply:
Ah, I see.

I think, it makes sense to have no entry function in that case, that would make it clear that there is no entrypoint.
lgc (middle-end) and llvm should be fine without an entrypoint, the shader stage is stored in function metadata in lgc and in the calling convention in llvm.
As far as I can see, the parts that need to be changed are

1. the SPIR-V execution model: I guess we can put this onto every function instead of only on entrypoints, similar to how it is done in lgc
2. lowering globals currently inserts code in the entrypoint: I think for non-entrypoint modules, this needs to be handled differently anyway. Maybe by moving the global into function arguments?

Reply:
our lowering pass assume the entry point a lot place in code, to rid of this assumption, i am afraid there are more than a few places to modify and adding code complexity.
It's about trade-off.  to add dummy entry point , which you can argue it is a little bit hack, is simple and easy without adding a lot of technical liability.